04/scala/src/main/scala/CollectionsLeetCodeOptional.scala

```diff
@@ -0,0 +1,126 @@
+  // https://leetcode.com/problems/maximum-nesting-depth-of-the-parentheses/
+  //
+  // A string is a valid parentheses string (denoted VPS) if it meets one of the following:
+  //   * It is an empty string "", or a single character not equal to "(" or ")",
+  //   * It can be written as AB (A concatenated with B), where A and B are VPS's, or
+  //   * It can be written as (A), where A is a VPS.
+  // We can similarly define the nesting depth depth(S) of any VPS S as follows:
+  //   * depth("") = 0
+  //   * depth(C) = 0, where C is a string with a single character not equal to "(" or ")".
+  //   * depth(A + B) = max(depth(A), depth(B)), where A and B are VPS's.
+  //   * depth("(" + A + ")") = 1 + depth(A), where A is a VPS.
+  // For example, "", "()()", and "()(()())" are VPS's (with nesting depths 0, 1, and 2), and ")(" and "(()" are not VPS's.
+  // Given a VPS represented as string s, return the nesting depth of s.
+  //
+  // Example 1
+  // Input: s = "(1+(2*3)+((8)/4))+1"
+  // Output: 3
+  // Explanation: Digit 8 is inside of 3 nested parentheses in the string.
+  //
+  // Example 2
+  // Input: s = "(1)+((2))+(((3)))"
+  // Output: 3
+  //
+  // Example 3
+  // Input: s = "1+(2*3)/(2-1)"
+  // Output: 1
+  //
+  // Example 4
+  // Input: s = "1"
+  // Output: 0
+
+  def maxDepth(s: String): Int =
+    s.foldLeft(0, 0) { (state, c) =>
+      val (max, cur) = state
+      c match {
+        case '(' => (Math.max(max, cur + 1), cur + 1)
+        case ')' => (max, cur - 1)
+        case _ => (max, cur)
+      }
+    } match {
+      case (x, 0) => x
+      case _ => 0
+    }
+
+  assert(maxDepth("(1+(2*3)+((8)/4))+1") == 3)
+  assert(maxDepth("(1)+((2))+(((3)))") == 3)
+  assert(maxDepth("1+(2*3)/(2-1)") == 1)
+  assert(maxDepth("1") == 0)
+
+  // https://leetcode.com/problems/split-a-string-in-balanced-strings/
+  //
+  // Balanced strings are those that have an equal quantity of 'L' and 'R' characters.
+  // Given a balanced string s, split it in the maximum amount of balanced strings.
+  // Return the maximum amount of split balanced strings.
+  //
+  // Example 1
+  // Input: s = "RLRRLLRLRL"
+  // Output: 4
+  // Explanation: s can be split into "RL", "RRLL", "RL", "RL", each substring contains same number of 'L' and 'R'.
+  //
+  // Example 2
+  // Input: s = "RLLLLRRRLR"
+  // Output: 3
+  // Explanation: s can be split into "RL", "LLLRRR", "LR", each substring contains same number of 'L' and 'R'.
+  //
+  // Example 3
+  // Input: s = "LLLLRRRR"
+  // Output: 1
+  // Explanation: s can be split into "LLLLRRRR".
+  //
+  // Example 4
+  // Input: s = "RLRRRLLRLL"
+  // Output: 2
+  // Explanation: s can be split into "RL", "RRRLLRLL", since each substring contains an equal number of 'L' and 'R'
+
+  def balancedStringSplit(s: String): Int = {
+    def toState(count: Int, cur: Int): (Int, Int) =
+      if (cur == 0) (count + 1, cur) else (count, cur)
+
+      s.foldLeft(0, 0) { (state, c) => {
```

Comment:
This last parenthesis is from the pair which can be omitted and usually is omitted
You could also do `{ case ((count, cur), c) =>` and omit the next line